zeroex/orderwatch/order_watcher.go

```diff
@@ -319,318 +336,91 @@ func (w *Watcher) handleOrderExpirations(validationBlock *types.MiniHeader, orde
+	w.recentlyValidatedOrdersMu.Unlock()
+
+	var firstRevalidationBlockNumber *big.Int
+	revalidationBlockToOrder := map[*big.Int][]*types.OrderWithMetadata{}
+	for _, recentlyValidatedOrder := range recentlyValidatedOrders {
+		previousValidationBlockNumber := recentlyValidatedOrder.LastValidatedBlockNumber
+		// If the oldestBlock in the list of block events is greater then
+		// the last validated block of the recently validated orders, we
+		// are missing block events for this order.
+		if firstRevalidationBlockNumber == nil || previousValidationBlockNumber.Cmp(firstRevalidationBlockNumber) == -1 {
+			firstRevalidationBlockNumber = previousValidationBlockNumber
+		}
+		if firstBlockFromEvents.Number.Cmp(previousValidationBlockNumber) == -1 {
+			continue
+		}
+		// If the previous validation block of the order is a predecessor
+		// of the oldest block in the blockwatcher, we must revalidate the
+		// order because we may be missing relevant block events.
+		if firstBlockInDB.Number.Cmp(previousValidationBlockNumber) == 1 {
+			orderHashToDBOrder[recentlyValidatedOrder.Hash] = recentlyValidatedOrder
+			orderHashToEvents[recentlyValidatedOrder.Hash] = []*zeroex.ContractEvent{}
+		}
+		revalidationBlockToOrder[previousValidationBlockNumber] = append(
+			revalidationBlockToOrder[previousValidationBlockNumber],
+			recentlyValidatedOrder,
+		)
+	}
 
-			case "ExchangeCancelUpToEvent":
-				var exchangeCancelUpToEvent decoder.ExchangeCancelUpToEvent
-				err = w.eventDecoder.Decode(log, &exchangeCancelUpToEvent)
-				if err != nil {
-					if isNonCritical := w.checkDecodeErr(err, eventType); isNonCritical {
-						continue
-					}
-					return err
-				}
-				contractEvent.Parameters = exchangeCancelUpToEvent
-				cancelledOrders, err := w.db.FindOrders(&db.OrderQuery{
-					Filters: []db.OrderFilter{
-						{
-							Field: db.OFMakerAddress,
-							Kind:  db.Equal,
-							Value: exchangeCancelUpToEvent.MakerAddress,
-						},
-						{
-							Field: db.OFSalt,
-							Kind:  db.LessOrEqual,
-							Value: exchangeCancelUpToEvent.OrderEpoch,
-						},
-					},
-				})
-				if err != nil {
-					logger.WithFields(logger.Fields{
-						"error": err.Error(),
-					}).Error("unexpected query error encountered")
-					return err
-				}
-				orders = append(orders, cancelledOrders...)
+	revalidationMiniHeaders, err := w.db.FindMiniHeaders(&db.MiniHeaderQuery{
+		Filters: []db.MiniHeaderFilter{
+			{
+				Field: db.MFNumber,
+				Kind:  db.Less,
+				Value: firstBlockFromEvents.Number,
+			},
+			{
+				Field: db.MFNumber,
+				Kind:  db.Greater,
+				Value: firstRevalidationBlockNumber,
+			},
+		},
+		Sort: []db.MiniHeaderSort{
+			{
+				Field:     db.MFNumber,
+				Direction: db.Ascending,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
 
-			default:
-				logger.WithFields(logger.Fields{
-					"eventType": eventType,
-					"log":       log,
-				}).Error("unknown eventType encountered")
+	eventFilter := map[common.Hash]bool{}
```

Comment:
The `eventFilter` is used to filter for only the orders in the map. It may not be the best name, and it's an efficient way of doing things (as you point out in another comment). Refactoring to avoid this does not seem trivial though, so I'll just add these fixes for now.